plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/p3order/LayerSweepCrossingMinimizer.java

```diff
@@ -226,19 +230,83 @@ private int minimizeCrossingsWithCounter(final GraphInfoHolder gData) {
 
         return oldNumberOfCrossings;
     }
+    
+    /**
+     * Compares all nodes in a each layer and counts how often they are not in model order.
+     * This requires that the {@code SortByInputModelProcessor} ran previously.
+     * @param layers layers to check
+     * @param strategy the ordering strategy to compare the nodes
+     * @return The number of model order conflicts
+     */
+    private int countModelOrderNodeChanges(final LNode[][] layers, final OrderingStrategy strategy) {
+        int previousLayer = -1;
+        int wrongModelOrder = 0;
+        for (LNode[] layer : layers) {
+            ModelOrderNodeComparator comp = new ModelOrderNodeComparator(
+                    previousLayer == -1 ? layers[0] : layers[previousLayer], strategy, DummyNodeStrategy.EQUAL);
+            for (int i = 0; i < layer.length; i++) {
+                for (int j = i + 1; j < layer.length; j++) {
+                    if (comp.compare(layer[i], layer[j]) > 0) {
+                        wrongModelOrder++;
+                    }
+                }
+            }
+            previousLayer++;
+        }
+        return wrongModelOrder;
+    }
+
+    /**
+     * Compares all ports in a each layer and counts how often they are not in model order.
+     * This requires that the {@code SortByInputModelProcessor} ran previously.
+     * @param layers layers to check
+     * @return The number of model order conflicts
+     */
+    private int countModelOrderPortChanges(final LNode[][] layers) {
+        int previousLayer = -1;
+        int wrongModelOrder = 0;
+        for (LNode[] layer : layers) {
+            for (LNode lNode : layer) {
+                Comparator<LPort> comp = new ModelOrderPortComparator(
+                        previousLayer == -1 ? layers[0] : layers[previousLayer],
+                                lNode.getProperty(InternalProperties.TARGET_NODE_MODEL_ORDER));
+                for (int i = 0; i < lNode.getPorts().size(); i++) {
+                    for (int j = i + 1; j < lNode.getPorts().size(); j++) {
+                        if (comp.compare(lNode.getPorts().get(i), lNode.getPorts().get(j)) > 0) {
+                            wrongModelOrder++;
+                        }
+                    }
+                }
+            }
+            previousLayer++;
+        }
+        return wrongModelOrder;
+    }
 
     /*
      * We only need to count crossings below the current graph and also only if they are marked as to be processed
      * hierarchically.
      */
-    private int countCurrentNumberOfCrossings(final GraphInfoHolder currentGraph) {
-        int totalCrossings = 0;
+    private double countCurrentNumberOfCrossings(final GraphInfoHolder currentGraph) {
```

Comment:
Do you see a chance to preserve the `int` here? Two reasons:
1) int feels more natural for a _count_
2) int avoids having to think about floating point equality

Reply:
I do not see how I could do this easily. It can happen that the node/port order influence are a fraction of the crossing influence and the other way around. I could of course calculate the factor to multiply each influence such that we would always have whole numbers, but I guess then we would also run in overflow problems or do you think that this will not occur?

Reply:
I share your concern regarding overflows and don't think scaling is a viable solution. 

However, I feel that the change `int` -> `double` and adding further cases to the "old" crossing minimizations could introduce hidden bugs.
I'd rather prefer a completely separate control flow for the interactive/preserving part, ie.
```
process(...) {
  if (preservingLayout)
     newMethod()
  else 
     existingMethods()
}
```

Reply:
@uruuru I refactored the node and port order weighting into separate methods. However, I think this might complicate things since it produces duplicate code (because of the three methods using double instead of int).